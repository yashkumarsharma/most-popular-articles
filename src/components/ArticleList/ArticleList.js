import PropTypes from 'prop-types'
import Skeleton from 'react-loading-skeleton'

import { getThumbnailImage, formatDate } from '../../utils'

const ArticleList = (props) => {
  const {
    isLoading = false,
    articles = [],
    onClick = () => {},
  } = props

  const renderList = (article = {}) => {
    const { title, media, id, published_date } = article
    const image = getThumbnailImage(media)

    return (
      <button
        key={id}
        onClick={() => onClick(id)}
        className='list-card border rounded-lg p-4 hover:border-blue-500 flex items-center bg-gray-200 mb-2 h-20 w-full'
        data-testid='list-card'
      >
        <div className='mr-4 flex-shrink-0'>
          {!!image && (
            <img className='w-20 h-20 object-cover' src={image} alt={title} />
          )}
        </div>
        <div>
          <p className='font-semibold line-clamp-2'>{title}</p>
          <div className='text-sm text-gray-500'>
            <p data-testid='date'>{formatDate(published_date)}</p>
          </div>
        </div>
      </button>
    )
  }

  return (
    <div className='md:w-2/5 md:max-h-full m-10'>
      <div>
        <h2 className='text-lg font-semibold mb-4'> Article List </h2>
        {!!isLoading && (
          <Skeleton
            count={10}
            key={`list-loader`}
            data-testid='list-loader'
            className='border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full'
          />
        )}
        {!isLoading && articles.map(renderList)}
      </div>
    </div>
  )
}

ArticleList.propTypes = {
  isLoading: PropTypes.bool,
  articles: PropTypes.arrayOf(PropTypes.shape({})),
  onClick: PropTypes.func,
}

export default ArticleList
