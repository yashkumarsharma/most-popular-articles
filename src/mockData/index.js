// eslint-disable-next-line import/prefer-default-export
export const mockArticles = [{
  'uri': 'nyt://article/7d22b056-2e3b-5778-b1c4-526f651778b9',
  'url': 'https://www.nytimes.com/2022/03/22/magazine/ethicist-teenage-sex-parenting.html',
  'id': 100000008260402,
  'asset_id': 100000008260402,
  'source': 'New York Times',
  'published_date': '2022-03-22',
  'updated': '2024-04-25 13:35:02',
  'section': 'Magazine',
  'subsection': '',
  'nytdsection': 'magazine',
  'adx_keywords': 'Ethics (Personal);Content Type: Service',
  'column': null,
  'byline': 'By Kwame Anthony Appiah',
  'type': 'Article',
  'title': 'My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?',
  'abstract': 'The magazine\'s Ethicist columnist on whether to keep your teenagerâ€™s.',
  'des_facet': [
    'Ethics (Personal)',
    'Content Type: Service',
  ],
  'org_facet': [],
  'per_facet': [],
  'geo_facet': [],
  'media': [
    {
      'type': 'image',
      'subtype': 'photo',
      'caption': '',
      'copyright': 'Illustration by Tomi Um',
      'approved_for_syndication': 0,
      'media-metadata': [
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-thumbStandard.jpg',
          'format': 'Standard Thumbnail',
          'height': 75,
          'width': 75,
        },
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo210.jpg',
          'format': 'mediumThreeByTwo210',
          'height': 140,
          'width': 210,
        },
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo440.jpg',
          'format': 'mediumThreeByTwo440',
          'height': 293,
          'width': 440,
        },
      ],
    },
  ],
  'eta_id': 0,
}]
