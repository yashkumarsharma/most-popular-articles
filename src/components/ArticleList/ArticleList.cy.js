import React from 'react'
import ArticleList from './ArticleList'
import { mockArticles } from './mockData'

describe('<ArticleList />', () => {
  it('renders', () => {
    // see: https://on.cypress.io/mounting-react
    cy.mount(<ArticleList articles={mockArticles} />)
  })
})