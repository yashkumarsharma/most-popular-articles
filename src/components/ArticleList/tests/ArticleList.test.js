import React from 'react'
import renderer from 'react-test-renderer'
import ArticleList from '../ArticleList'
import { mockArticles } from '../../../mockData'

describe('check ArticleList', () => {
  it('renders correctly when no articles passed', () => {
    const tree = renderer
      .create(<ArticleList />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when Component isLoading', () => {
    const tree = renderer
      .create(<ArticleList isLoading={true} articles={mockArticles} onClick={() => {}} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when articles are passed', () => {
    const tree = renderer
      .create(<ArticleList isLoading={false} articles={mockArticles} onClick={() => {}} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correct numer of li elements', () => {
    const tree = renderer
      .create(<ArticleList isLoading={false} articles={mockArticles} />)

    expect(tree.root.findAllByProps({ 'data-testid': 'list-card' }).length).toBe(mockArticles.length)
  })

  it('renders correct date on li elements', () => {
    const tree = renderer
      .create(<ArticleList isLoading={false} articles={mockArticles} />)

    expect(tree.root.findByProps({ 'data-testid': 'date' }).children).toStrictEqual([ '22 Mar 2022' ])
  })
})
