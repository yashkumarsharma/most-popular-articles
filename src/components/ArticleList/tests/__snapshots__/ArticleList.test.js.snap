// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check ArticleList renders correctly when Component isLoading 1`] = `
<div
  className="md:w-2/5 md:max-h-full m-10"
>
  <div>
    <h2
      className="text-lg font-semibold mb-4"
    >
       Article List 
    </h2>
    <span
      aria-busy={true}
      aria-live="polite"
    >
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
      <span
        className="react-loading-skeleton border rounded-lg p-4 hover:border-blue-500 flex-1 items-center bg-gray-200 mb-2 h-20 w-full"
        style={Object {}}
      >
        ‌
      </span>
      <br />
    </span>
  </div>
</div>
`;

exports[`check ArticleList renders correctly when articles are passed 1`] = `
<div
  className="md:w-2/5 md:max-h-full m-10"
>
  <div>
    <h2
      className="text-lg font-semibold mb-4"
    >
       Article List 
    </h2>
    <button
      className="list-card border rounded-lg p-4 hover:border-blue-500 flex items-center bg-gray-200 mb-2 h-20 w-full"
      data-testid="list-card"
      onClick={[Function]}
    >
      <div
        className="mr-4 flex-shrink-0"
      >
        <img
          alt="My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?"
          className="w-20 h-20 object-cover"
          src="https://static01.nyt.com/images/2022/03/27/magazine/27mag-ethicist/27mag-ethicist-thumbStandard.jpg"
        />
      </div>
      <div>
        <p
          className="font-semibold line-clamp-2"
        >
          My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?
        </p>
        <div
          className="text-sm text-gray-500"
        >
          <p
            data-testid="date"
          >
            22 Mar 2022
          </p>
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`check ArticleList renders correctly when no articles passed 1`] = `
<div
  className="md:w-2/5 md:max-h-full m-10"
>
  <div>
    <h2
      className="text-lg font-semibold mb-4"
    >
       Article List 
    </h2>
  </div>
</div>
`;
